utils_test.py

```diff
@@ -246,3 +247,117 @@ def test_get_hashable_value(self):
+        self.assertEqual(call_counter[nonempty_kwargs], 0)
+
+        count_calls()
+        self.assertEqual(call_counter[empty_kwargs], 1)
+        self.assertEqual(call_counter[nonempty_kwargs], 0)
+
+        count_calls()
+        count_calls(kwarg=0)
+        self.assertEqual(call_counter[empty_kwargs], 1)
+        self.assertEqual(call_counter[nonempty_kwargs], 1)
+
+    def test_memoize_with_kwargs_using_default_values(self):
+        call_counter = collections.Counter()
+
+        @utils.memoize
+        def count_calls(kwarg=0):
+            """Counts calls made with given kwargs."""
+            call_counter[kwarg] += 1
+
+        self.assertEqual(call_counter[0], 0)
+        count_calls()
+        self.assertEqual(call_counter[0], 1)
+        count_calls(kwarg=0)
+        self.assertEqual(call_counter[0], 1)
+
+    def test_memoize_with_methods(self):
+        class CallCounter(object):
+            """Counts calls made to an instance."""
+            def __init__(self):
+                self.count = 0
+
+            @utils.memoize
+            def __call__(self):
+                self.count += 1
+
+        call_counter_a, call_counter_b = CallCounter(), CallCounter()
+        self.assertEqual(call_counter_a.count, 0)
+        self.assertEqual(call_counter_b.count, 0)
+
+        call_counter_a()
+        self.assertEqual(call_counter_a.count, 1)
+        self.assertEqual(call_counter_b.count, 0)
+
+        call_counter_a()
+        call_counter_b()
+        self.assertEqual(call_counter_a.count, 1)
+        self.assertEqual(call_counter_b.count, 1)
+
+    def test_memoize_with_classmethods(self):
+        class GoodCallCounter(object):
+            """Counts calls made to the class."""
+            count = 0
+
+            @classmethod
+            @utils.memoize
+            def method_decorated_by_memoize_before_classmethod(cls):
+                """memoize is called first so this def will work properly."""
+                cls.count += 1
+
+        call_counter_a, call_counter_b = GoodCallCounter(), GoodCallCounter()
+        self.assertEqual(GoodCallCounter.count, 0)
+
+        call_counter_a.method_decorated_by_memoize_before_classmethod()
+        self.assertEqual(GoodCallCounter.count, 1)
+
+        call_counter_a.method_decorated_by_memoize_before_classmethod()
+        call_counter_b.method_decorated_by_memoize_before_classmethod()
+        self.assertEqual(GoodCallCounter.count, 1)
+
+        with self.assertRaisesRegexp(TypeError, 'not a Python function'):
+            class BadCallCounter(object):  # pylint: disable=unused-variable
+                """Counts calls made to the class."""
+                count = 0
+
+                @utils.memoize
+                @classmethod
+                def method_decorated_by_classmethod_before_memoize(cls):
+                    """classmethods are not real functions so trying to memoize
+                    them will raise a TypeError.
+                    """
```

Comment:
Add pass statement below this line

Reply:
I wanted to, but it becomes unreachable code. I'd prefer to leave it out since this is still valid Python :(